### ISSUE TITLE
Not connecting to KeepassXC

### ISSUE THREAD

**Issue by ericblank** [varjolintu/keepassxc-browser#38]
## Expected Behavior
Connect to KeepassXC

## Current Behavior
Running KeepassXC with Keepass HTTP server, firefox addon does not connect

## Steps to Reproduce (for bugs)

1. Go to Settings, Connected Databases
2. Click connect
3. Nothing happens
4.

## General Info
KeePassXC fork - 2.2.2
keepassxc-browser -  0.4.2

Operating system: Ubuntu (Xubuntu)
Browser: Firefox Quantum
Proxy used: NO

**Comment by varjolintu** [comment]
The browser extension is not KeePassHTTP compatible. You must enable the `Browser extension with native messaging` in the KeePassXC settings.

**Comment by ericblank** [comment]
I've disabled KeepassHTTP and enabled Native Messaging, still no effect

**Comment by varjolintu** [comment]
Have you read the [wiki](https://github.com/varjolintu/keepassxc-browser/wiki/Connecting-the-database-with-keepassxc-browser) for troubleshooting? If you are not using a proxy you must have KeePassXC closed before starting the browser. This is because browser starts the KeePassXC process. If you like to use KeePassXC just like with KeePassHTTP, use the proxy feature. It can be enabled in the Advanced tab of the _Browser extension with native messaging_ settings.

**Comment by ericblank** [comment]
I followed the wiki steps. KeepassXC is not launching. I'm using the appimage and a symlink to /usr/bin/keepassxc

**Comment by varjolintu** [comment]
What does the JavaScript console say if you debug the browser extension? The issue might be that the .json holds an incorrect path (this can happen when using AppImage) to the keepassxc binary. Could you check those two?

**Comment by ericblank** [comment]
I've opened the js console but nothing comes up, but I might be doing it wrong. it's just f12 -> console?

**Comment by varjolintu** [comment]
Go to `about:Debugging`, check `Enable add-on debugging`. Then choose keepassxc-browser from the list.

**Comment by flxchn** [comment]
the JavaScript console say

Webconsole context has changed
Connecting to native messaging host com.varjolintu.keepassxc_browser  keepass.js:855:2
Failed to connect: Unknown error  keepass.js:831:2
Error 5: Timeout or not connected to KeePassXC  keepass.js:914:2
Connecting to native messaging host com.varjolintu.keepassxc_browser  keepass.js:855:2
Failed to connect: Unknown error  keepass.js:831:2
Connecting to native messaging host com.varjolintu.keepassxc_browser  keepass.js:855:2
Failed to connect: Unknown error  keepass.js:831:2
Connecting to native messaging host com.varjolintu.keepassxc_browser  keepass.js:855:2
Failed to connect: Unknown error  keepass.js:831:2
Error 5: Timeout or not connected to KeePassXC keepass.js:914:2
The character encoding of a framed document was not declared. The document may appear different if viewed without the document framing it.  hscv
Error 5: Timeout or not connected to KeePassXC keepass.js:914:2
Attempt to set a forbidden header was denied: Connection 2767292510-lcs_client_bin.js:99:385
Error 5: Timeout or not connected to KeePassXC keepass.js:914:2
Webconsole context has changed
Error: Promised response from onMessage listener at @moz-extension://664c660a-e78c-43a8-8a8e-f483121afb58/background/init.js:92:1 went out of scope undefined

And Both browsers addons(Firefox/Chromium)  not work

**Comment by varjolintu** [comment]
Next step, check the path variable in these files and see where it points to:
`~/.mozilla/native-messaging-hosts/com.varjolintu.keepassxc_browser.json'
`~/.config/chromium/NativeMessagingHosts/com.varjolintu.keepassxc_browser.json`

Because you are using an AppImage, you can also try to enable the proxy with binary updating checked in the Advanced settings tab. Still, there might be issues with Unix sockets and AppImage and they might not work without some extra configuring.

**Comment by ericblank** [comment]
I'm getting the same Error 5 timeout, I'll try the path variables

**Comment by ericblank** [comment]
I changed the path to my symlink, keepassxc launches with the browser now. 

Now it says I have an outdated version (assuming it's because it's a fork?) and it says:

 keepassxc-browser has encountered an error:
Cannot decrypt message

**Comment by varjolintu** [comment]
Yes, the outdated version message happens because of this.

Did you made a symlink from /usr/bin/keepassxc to the .AppImage or what method did you actually use? I could try to reproduce the exact problem.

**Comment by ericblank** [comment]
Yes that's what I did

On December 10, 2017 10:31:18 PM PST, "Sami Vänttinen" <notifications@github.com> wrote:
>Yes, the outdated version message happens because of this.
>
>Did you made a symlink from /usr/bin/keepassxc to the .AppImage or what
>method did you actually use? I could try to reproduce the exact
>problem.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/varjolintu/keepassxc-browser/issues/38#issuecomment-350633982

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by varjolintu** [comment]
The symlink needs to be pointing to the exact executable binary file, not to an .AppImage itself.

**Comment by gamusin** [comment]
I'm having the same problem cannot connect followed every step in the wiki
**plugin debug:**
`TypeError: invalid 'in' operand browser[Learn More]  tabbrowser.xml:2346:1
BindInt32Parameter is deprecated and will be removed soon.  (unknown)
BindStringParameter is deprecated and will be removed soon. (unknown)
BindInt32Parameter is deprecated and will be removed soon.  (unknown)
this.browser is null ExtensionTabs.jsm:305
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIChannel.contentType]  converter-observer.js:91
BindInt32Parameter is deprecated and will be removed soon. (unknown)
SyntaxError: The URI is malformed.  (unknown)
TypeError: invalid 'in' operand keepass.keyRing[Learn More]  keepass.js:682:4
Connecting to native messaging host org.keepassxc.keepassxc_browser  keepass.js:861:2
Unix error 2 during operation open on file /sda3/home/.cache/waterfox/ahhdz7lj.default/directoryLinks.json (No existe el fichero o el directorio)  DirectoryLinksProvider.jsm:254
This extension does not have permission to use native application org.keepassxc.keepassxc_browser (or the application is not installed)  NativeMessaging.jsm:216
Failed to connect: Unknown error  keepass.js:837:2
Error 5: Timeout or not connected to KeePassXC  keepass.js:920:2
BindInt32Parameter is deprecated and will be removed soon.  (unknown)
Elemento clave no disponible en algunas disposiciones de teclado: key=“r” modifiers=“accel,alt” id=“key_toggleReaderMode”  browser.xul
Elemento clave no disponible en algunas disposiciones de teclado: key=“i” modifiers=“accel,alt,shift” id=“key_browserToolbox”  browser.xul
Content Security Policy: Ignorando ‘x-frame-options’ por la directiva ‘frame-ancestors’.  (unknown)
Content Security Policy: La configuración de la página bloqueó la carga de un recurso en self ("script-src https://assets-cdn.github.com"). Source: (function e(t,n,r){function s(o,u){if(!n....  38:1
Sync encountered an error - see about:sync-log for the log file.  policies.js:772
Warning: ‘nsIOService::NewChannel()’ obsoleto, use ‘nsIOService::NewChannel2()’`

**Comment by gamusin** [comment]
I'm using waterfox over here.

**Comment by varjolintu** [comment]
@gamusin Make sure you use the development build of the official KeePassXC with browser extension version 0.4.8. If you already use them make sure to install the browser support again from the KeePassXC Settings page. This needs to be done because the native messaging host name has changed. The line `This extension does not have permission to use native application org.keepassxc.keepassxc_browser (or the application is not installed)` clearly shows the issue is related to this.

**Comment by gamusin** [comment]
hmm ok thanks I'm using a git version I found on a opensuse repo not sure of what date. And what about the line that say it cannot fin directoryLinks.json that's a windows file I think I cannot find it in my linux system

**Comment by gamusin** [comment]
development version you mean git version or another version. Thanks for the answer.

**Comment by varjolintu** [comment]
@gamusin Git version yes. From the development branch. AFAIK there's beta/nightly builds coming soon to the official repository.

**Comment by gamusin** [comment]
no just recompiled everything it doesn't work but it seems the problem is only with waterfox with firefox 57 it works. It isn't exactly a keepassxc problem.

**Comment by gamusin** [comment]
**new debug log** (after recompiling)
`Connecting to native messaging host org.keepassxc.keepassxc_browser  keepass.js:861:2
Failed to connect: Unknown error  keepass.js:837:2
BindInt32Parameter is deprecated and will be removed soon.  (unknown)
Sync encountered an error - see about:sync-log for the log file.  policies.js:772
Content Security Policy: Ignorando ‘x-frame-options’ por la directiva ‘frame-ancestors’. (unknown)
Error 5: Timeout or not connected to KeePassXC keepass.js:920:2
Elemento clave no disponible en algunas disposiciones de teclado: key=“r” modifiers=“accel,alt” id=“key_toggleReaderMode”  browser.xul
Elemento clave no disponible en algunas disposiciones de teclado: key=“i” modifiers=“accel,alt,shift” id=“key_browserToolbox”  browser.xul
Content Security Policy: La configuración de la página bloqueó la carga de un recurso en self ("script-src https://assets-cdn.github.com"). Source: (function e(t,n,r){function s(o,u){if(!n....  38:1
Error 5: Timeout or not connected to KeePassXC  keepass.js:920:2
Connecting to native messaging host org.keepassxc.keepassxc_browser  keepass.js:861:2
This extension does not have permission to use native application org.keepassxc.keepassxc_browser (or the application is not installed)  NativeMessaging.jsm:216
Failed to connect: Unknown error  keepass.js:837:2
Error 5: Timeout or not connected to KeePassXC  keepass.js:920:2
TypeError: keepass.nativePort is null[Learn More]  keepass.js:90:7
uncaught exception: false  (unknown)
Use of nsIFile in content process is deprecated.  loader.js:247:15
TelemetryStopwatch: requesting elapsed time for nonexisting stopwatch. Histogram: "PLACES_AUTOCOMPLETE_1ST_RESULT_TIME_MS", key: "null"  TelemetryStopwatch.jsm:363
Warning: ‘nsIOService::NewChannel()’ obsoleto, use ‘nsIOService::NewChannel2()’`

**Comment by varjolintu** [comment]
I'm not sure where Waterfox tries to find the native messaging JSON scripts. Please check if the script files needs to be located under .waterfox. Copying the native messaging folder from under .mozilla could work.

**Comment by gamusin** [comment]
yes it worked!! thanks very much just one question sorry to bother you. Why does waterfox/firefox keeps asking to save the passwords even though they are already stored in keepass, ipassifox worked differently.

**Comment by gamusin** [comment]
ipassifox integrated differently it replaced the firefox password manager if a password was already stored in keepass it would ask to update it for example but not to save the password again twice.

**Comment by varjolintu** [comment]
WebExtension doesn't have access to the internal password manager, so you better disable it. That should solve the issue.

**Comment by cdavila58** [comment]
I am getting the same error. I recently installed keepassxc v. 2.2.4 on Ubuntu and the Firefox extension KeePassXC-Browser Version: 0.5.2. When I click on the extension icon, I get:

" KeePassXC-Browser has encountered an error:

Timeout or not connected to KeePassXC "

when I click on the blue "Reload" button, I get the following on the Javascript console:

![image](https://user-images.githubusercontent.com/17222917/36357170-b645bd2e-14bf-11e8-9e9c-941a541368e5.png)
I followed all the directions in the documentation:

![image](https://user-images.githubusercontent.com/17222917/36357276-567f156e-14c1-11e8-8d9a-f06e9471c797.png)

![image](https://user-images.githubusercontent.com/17222917/36357283-67c3f6aa-14c1-11e8-9f62-d46d451d81f0.png)
 and no luck.

**Comment by varjolintu** [comment]
@cdavila58 Currently this needs a development build of KeePassXC. If you don't want to build it by yourself you can wait for a beta release that is coming soon.

[This page](https://keepassxc.org/docs/keepassxc-browser-migration/) is the updated guide for the new browser integration.

**Comment by Ovsyanka** [comment]
I got the same issue (on the windows system, though).

@varjolintu, Thank you for the explanation.

But it looks quite odd that actual version of keepassxc-browser don't work with the actual version of KeepasXC. How it happened at all?

**Comment by varjolintu** [comment]
@Ovsyanka This is because of a transition period that's going on. KeePassXC is finalizing a new release, and my own KeePassXC fork used extension id's I created on my own. So now we had to publish new extension with new ID's which are related to KeePassXC Team and not just me. Also, because KeePassXC has restrictions about what extensions are allowed to connect to it, that list has changed to allow only the new official ones. We didn't want to include any older or unnecessary extension ID's to the allowed list of extensions.

I know this kind of transition period can cause some problems with users that had already downloaded my KeePassXC fork while extension have been updated to use the new id's.

For now, if you want to use the browser extensions you must compile the KeePassXC snapshot manually from the official repository. Releases from my KeePassXC fork page are not recommended, and I'll delete them very soon.

Hope this explains the mess a little.

**Comment by Ovsyanka** [comment]
@varjolintu, thank you for the detailed explanation. Initially I didn't know about your fork at all. I just installed official (keepassxreboot/keepassxc) KeepassXC and official firefox addon [keepassxc-browser](https://addons.mozilla.org/ru/firefox/addon/keepassxc-browser/). I investigate the situation and realised, that the wrong thing is that keepassxc-browser page links to the official KeepassXC, but in fact it works with dev version of KeepassXC or your fork. Am I right? In that case seems right for me to write about this in the addon page and point to the official KeepassXC only when it will be in the current release.

It is on you and other developers to decide what to do, I just share my vision.

Thank you for your work!

**Comment by varjolintu** [comment]
@Ovsyanka Yes, you are correct. Currently it's best to use a dev version of the official KeePassXC and not my fork.

The extension description held the info but other devs removed it, so I'm not wondering if it causes some confusion. Some have even approached the devs and asked if the extension is even legit.

Thank you for the feedback!

**Comment by cbh2** [comment]
Good evening. I have followed all comments and instructions in how to connect the KeePassXC with the KeePassXC Browser and still I cannot connect the two.

I have a Linux 64 bit platform, more exactly Linux Mint 18.3 I use Opera as my main browser as Firefox have problems with some local websites. Anyway Opera can incorperate Chrome extensions and then work the KeePassXC Browser through that.

The problem is that I can still only use the KeePassXC via the KeePassHttp. If I untick the box "Activate KeePassHttp server" in Legacy Browser Integration I have no connection at all.

I am sorry to say, but all the explanation in GitHub and other places how to connect the two mentioned in the first part is only for programming nerds. About the proxy and other things mentioned is not to understand unless you are an expert.

I have humble question. Would it be possible to make a step by step instruction for "dummies" how to connect KeePassXC with KeePassXC Browser for Linux users. I have installed everything mentioned in the guides to start with. But then when you talk about proxy and further I am lost, as well with at lot of others I presume.

Thank you in advance Chris

**Comment by cbh2** [comment]
OK, let us take it for Firefox, which I also have. Please explain the the connection step by step. Do not take for granted that I know all about proxies, custom proxy locations etc.

The official guide is a bit better, although it does not solve my problem. I have tried also in Firefox.

**Comment by varjolintu** [comment]
@cbh2 So enabling the Browser Integration support and checking Firefox still doesn't work? [Here](https://github.com/keepassxreboot/keepassxc/issues/1559#issuecomment-369508900) is a bit nerdy troubleshooting that could help. Please check those points.

Don't worry about the proxy stuff. To make it clear, KeePassXC-Browser launches a `keepassxc-proxy` binary that transfers the messages between KeePassXC and the browser extension. Usually you don't have to pay attention to these extra configuration stuff.

**Comment by cbh2** [comment]
Thanks for the explanations. Those have got me a bit closer to understand the possible fault. I can localise the ~/.config/chromium/ but in there is no NativeMessagingHosts. This is in my Linux Mint 18.3 system. So it seams obvious that something vital is missing.

I have no problems with KeePassXC working together with the KeePassHelper, it works fine, but there I use the KeePassHttp extension. I am just worried how much longer I can use that system as I cannot get the KeePassXC Browser working correctly.

**Comment by varjolintu** [comment]
@cbh2 So you mentioned you are using Firefox but you're now looking native messaging script for Chromium? I'm confused. Do you have the browser support enabled for your brower in the KeePassXC settings?

**Comment by cbh2** [comment]
Sorry Varjolintu, I am using both to see if I can get the browser extension working in any of them. I would prefer Chromium though. Yes I have the browser support enabled.

**Comment by varjolintu** [comment]
Enabling the Chromium checkbox should create the necessary .json file under ~/.config/chromium/NativeMessagingHost. If the checkbox really doesn't work for some reason (it should display an error when saving the file fails) you can try to create the org.keepassxc.keepassxc_browser.json file manually with the following content:

```
{
    "allowed_origins": [
        "chrome-extension://oboonakemofpalcgghocfoadofidjkkk/"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "<path to keepassxc_proxy>",
    "type": "stdio"
}
```

**Comment by varjolintu** [comment]
Yes, so that file should be `~/.config/chromium/NativeMessagingHost/org.keepassxc.keepassxc_browser.json`.

You don't need to write anything to the custom proxy path.

Does KeePassXC create those files for some other browser? You can check the paths from the troubleshooting guide.

**Comment by cbh2** [comment]
"You don't need to write anything to the custom proxy path." That I understand, but in the script itself between the "<" and the ">" there should be a path written. 
"path": "<path to keepassxc_proxy>", does not point to anything. Where in the system do I have the keepassxc proxy?

**Comment by varjolintu** [comment]
Oh sorry, you meant that. `keepassxc_proxy` should be in the same directory where `keepassxc` binary is.

**Comment by cbh2** [comment]
A new comment in the above case.
I was asking you the path in Linux to where the keepassproxy was located so I could insert it in the "path" below I have your script and below that the script in the file KeePassXC did. See the path pointing to /usr/bin/keepassxc-proxy, that was the text I was asking you about.

{
    "allowed_origins": [
        "chrome-extension://oboonakemofpalcgghocfoadofidjkkk/"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "<path to keepassxc_proxy>",
    "type": "stdio"
}

------------------------------------------------------------------------------------------------------------

{
    "allowed_origins": [
        "chrome-extension://iopaggbpplllidnfmcghoonnokmjoicf/",
        "chrome-extension://oboonakemofpalcgghocfoadofidjkkk/"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "/usr/bin/keepassxc-proxy",          **< That was the path I asked you to describe**
    "type": "stdio"
}

**Comment by varjolintu** [comment]
@cbh2 Glad you got it finally fixed!

**Comment by rojagit** [comment]
I keep getting this error 
![image](https://user-images.githubusercontent.com/4193031/42967469-5796d7f6-8b6e-11e8-87f5-686f4ca73113.png)

**Comment by varjolintu** [comment]
@rojagit Please read the whole issue thread. You can try the things suggested before.

**Comment by Christos11** [comment]
Hello,

I can't get the firefox addon to connect to my -otherwise working fine- KeePassXC installation. I installed KeePassXC on my linux mint machine via the software manager. My debug info is as follows:

> KeePassXC - Version 2.3.3
> Revision: 0a155d8
> Distribution: Flatpak
> 
> Libraries:
> - Qt 5.10.1
> - libgcrypt 1.7.3
> 
> Operating system: Linux 4.15.0-29-generic
> CPU architecture: x86_64
> Kernel: linux 4.15.0-29-generic
> 
> Enabled extensions:
> - Auto-Type
> - Browser Integration
> - Legacy Browser Integration (KeePassHTTP)
> - SSH Agent
> - YubiKey

The contents of my `/home/pc/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json` file are:

```
{
    "allowed_extensions": [
        "keepassxc-browser@keepassxc.org"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "/var/lib/flatpak/exports/bin/org.keepassxc.KeePassXC",
    "type": "stdio"
}
```

The path `/var/lib/flatpak/exports/bin/org.keepassxc.KeePassXC` does not exist, and this is what the ffox debugging console reports too:

> File at path "/var/lib/flatpak/exports/bin/org.keepassxc.KeePassXC" does not exist, or is not executable

Please advise what I can do to get my browser talking to KPXC.

Thanks

**Comment by varjolintu** [comment]
@Christos11 Check this out: https://github.com/flathub/org.keepassxc.KeePassXC/issues/13
